File: ocaml/lib/rgssad_core.ml
(* RGSSAD v1 parser + payload decryptor.

   Used by BOTH RPG Maker XP (`.rgssad`) and VX (`.rgss2a`): the two share one
   on-disk format whose version byte is 0x01. (There is no RGSSAD "v2"; VX Ace's
   `.rgss3a` is v3 and lives in [Vxace].)

   The format is a stream, not a table. A single rolling key is seeded with
   0xDEADCAFE and mutated `key = (key*7 + 3) mod 2^32` after every 32-bit field
   it decrypts and after every individual filename byte. From offset 8 to EOF:

     name_len : u32                (XOR whole key, then advance)
     name     : name_len bytes     (each XOR key&0xff, advance per byte)
     size     : u32                (XOR whole key, then advance)
     data     : size bytes         (left in place; decrypted on demand with the
                                     key value captured right after [size])

   File data is decrypted separately by [decrypt_data]: the per-entry key
   (captured in [entry.key]) advances every 4 bytes and each cipher byte is
   XOR'd with the matching little-endian byte of the current key.

   Length fields are read into OCaml's 63-bit `int`, so a corrupt high-bit
   length stays a large POSITIVE value and is caught by the `pos + n > len`
   bounds checks (Truncated) rather than causing an out-of-bounds access. *)

type entry = { index : int; name : string; offset : int; size : int; key : int }
type parse_error = ShortHeader | BadMagic | BadVersion of int | Truncated

let magic_key = Crypto.magic_rgssad_prefix
let initial_key = 0xDEADCAFE
let u32 (x : int) : int = x land 0xFFFFFFFF
let advance (key : int) : int = u32 ((key * 7) + 3)

let read_u32_le (buf : bytes) (pos : int) : int =
  Char.code (Bytes.get buf pos)
  lor (Char.code (Bytes.get buf (pos + 1)) lsl 8)
  lor (Char.code (Bytes.get buf (pos + 2)) lsl 16)
  lor (Char.code (Bytes.get buf (pos + 3)) lsl 24)

let parse (buf : bytes) : (entry list * int, parse_error) result =
  let len = Bytes.length buf in
  if len < 8 then Error ShortHeader
  else if not (Crypto.starts_with magic_key buf) then Error BadMagic
  else if Char.code (Bytes.get buf 7) <> 0x01 then
    Error (BadVersion (Char.code (Bytes.get buf 7)))
  else begin
    let key = ref initial_key in
    let pos = ref 8 in
    let idx = ref 0 in
    let acc = ref [] in
    let keep = ref true in
    let malformed = ref false in
    while !keep do
      (* need a u32 name_len; otherwise we've reached the end of the stream *)
      if !pos + 4 > len then begin
        if !pos <> len then malformed := true;
        keep := false
      end
      else begin
        let name_len = read_u32_le buf !pos lxor !key in
        key := advance !key;
        pos := !pos + 4;
        if name_len < 0 || !pos + name_len > len then begin
          malformed := true;
          keep := false
        end
        else begin
          let name = Bytes.create name_len in
          for i = 0 to name_len - 1 do
            Bytes.set name i
              (Char.chr (Char.code (Bytes.get buf (!pos + i)) lxor (!key land 0xFF)));
            key := advance !key
          done;
          pos := !pos + name_len;
          if !pos + 4 > len then begin
            if !pos <> len then malformed := true;
            keep := false
          end
          else begin
            let size = read_u32_le buf !pos lxor !key in
            key := advance !key;
            pos := !pos + 4;
            let data_key = !key in
            if size < 0 || !pos + size > len then begin
              malformed := true;
              keep := false
            end
            else begin
              acc :=
                {
                  index = !idx;
                  name =
                    String.map
                      (fun c -> if c = '\\' then '/' else c)
                      (Bytes.to_string name);
                  offset = !pos;
                  size;
                  key = data_key;
                }
                :: !acc;
              incr idx;
              pos := !pos + size
            end
          end
        end
      end
    done;
    if !malformed then Error Truncated
    else match !acc with [] -> Error Truncated | l -> Ok (List.rev l, !pos)
  end

(** Extract raw (still-encrypted) payload bytes for one entry. *)
let read_entry (buf : bytes) (e : entry) : bytes =
  let start = e.offset in
  let len = Bytes.length buf in
  if start < 0 || start >= len || e.size <= 0 then Bytes.create 0
  else begin
    let size = min e.size (len - start) in
    Bytes.sub buf start size
  end

(** Decrypt an entry's payload: XOR each byte with the little-endian byte of the
    running key, advancing `key = key*7+3` every 4 bytes. *)
let decrypt_data (e : entry) (cipher : bytes) : bytes =
  let n = Bytes.length cipher in
  let out = Bytes.make n '\000' in
  let key = ref e.key in
  let j = ref 0 in
  for i = 0 to n - 1 do
    if !j = 4 then begin
      key := advance !key;
      j := 0
    end;
    Bytes.set out i
      (Char.chr (Char.code (Bytes.get cipher i) lxor ((!key lsr (8 * !j)) land 0xFF)));
    incr j
  done;
  out

File: fsharp/src/RpgmDecrypt.Core/RgssadCore.fs
namespace RpgmDecrypt.Core

/// Shared table-of-contents parser for RPG Maker XP (`0x01`) and VX (`0x02`)
/// `.rgssad` archives.
///
/// Both engines use the identical on-disk entry layout
///   `size(4) | offset(4) | name_len(4) | name(name_len)`
/// with filenames XOR-obfuscated by the 7-byte RGSSAD magic. They differ only
/// in the version byte and the end-of-table sentinel. Keeping the loop here —
/// rather than duplicated in Format.Xp.fs and Format.Vx.fs — means the bounds
/// and negative-`name_len` guard can never drift between the two formats again
/// (the original drift was AUDIT finding I-1).
///
/// VX Ace (`0x03`) is deliberately NOT handled here: it derives a master key
/// from a seed, encrypts every header field, and has per-entry payload
/// encryption — a genuinely different algorithm that lives in Format.VxAce.fs.
[<RequireQualifiedAccess>]
module RgssadCore =

    open System

    type Entry =
        { Index: int
          Name: string
          Offset: int64
          Size: int32 }

    type ParseError =
        | ShortHeader
        | BadMagic
        | BadVersion of byte
        | Truncated

    /// End-of-table rule — the only behavioural difference between XP and VX.
    type Sentinel =
        | NameLenZero      // XP: a name_len of 0 terminates the table
        | SizeAndNameZero  // VX: size = 0 && name_len = 0 terminates the table

    // 7-byte "RGSSAD\0" — explicit bytes; F# "RGSSAD\0" expands to 8 chars on
    // .NET 10, so we keep the array literal.
    let magicKey : byte[] =
        [| 0x52uy; 0x47uy; 0x53uy; 0x53uy; 0x41uy; 0x44uy; 0x00uy |]

    /// Filenames are XOR-obfuscated with the cycling magic prefix; trailing
    /// NULs (padding) are trimmed before decoding as UTF-8.
    let xorDecodeName (raw: byte[]) : string =
        if raw.Length = 0 then ""
        else
            let out = Array.zeroCreate<byte> raw.Length
            let keyLen = magicKey.Length
            for i = 0 to raw.Length - 1 do
                out.[i] <- raw.[i] ^^^ magicKey.[i % keyLen]
            let mutable endIdx = out.Length
            while endIdx > 0 && out.[endIdx - 1] = 0uy do endIdx <- endIdx - 1
            let slice = Array.zeroCreate<byte> endIdx
            Array.Copy(out, slice, endIdx)
            System.Text.Encoding.UTF8.GetString slice

    let private readU32LE (buf: byte[]) (pos: int) : uint32 =
        uint32 buf.[pos]
        ||| (uint32 buf.[pos + 1] <<< 8)
        ||| (uint32 buf.[pos + 2] <<< 16)
        ||| (uint32 buf.[pos + 3] <<< 24)

    /// Parse the entry table. Returns the entries plus the byte-offset one past
    /// the last record read. `version` is the expected byte at offset 7;
    /// `sentinel` is the format's end-of-table rule.
    let parse (version: byte) (sentinel: Sentinel) (buf: byte[])
        : Result<Entry list * int, ParseError> =
        if buf.Length < 8 then Error ShortHeader
        elif not (Crypto.startsWith magicKey buf) then Error BadMagic
        elif buf.[7] <> version then Error(BadVersion buf.[7])
        else
            let mutable pos = 8
            let mutable idx = 0
            let acc = ResizeArray<Entry>()
            let mutable keepGoing = true
            while keepGoing do
                if pos + 12 > buf.Length then keepGoing <- false
                else
                    let size    = readU32LE buf pos
                    pos <- pos + 4
                    let offset  = readU32LE buf pos
                    pos <- pos + 4
                    let nameLen = readU32LE buf pos
                    pos <- pos + 4
                    let nameLenInt = int nameLen
                    let isEnd =
                        match sentinel with
                        | NameLenZero     -> nameLen = 0u
                        | SizeAndNameZero -> nameLen = 0u && size = 0u
                    if isEnd then
                        keepGoing <- false
                    elif nameLenInt < 0 || pos + nameLenInt > buf.Length then
                        // nameLen wrapped negative (high bit set) or runs past
                        // the buffer; treat as a corrupt/truncated table rather
                        // than throwing inside Array.zeroCreate/Array.Copy.
                        // Clearing acc makes the post-loop count check report
                        // Truncated.
                        acc.Clear()
                        keepGoing <- false
                    else
                        let nameBytes = Array.zeroCreate<byte> nameLenInt
                        Array.Copy(buf, pos, nameBytes, 0, nameLenInt)
                        pos <- pos + nameLenInt
                        acc.Add
                            { Index = idx
                              Name = xorDecodeName nameBytes
                              Offset = int64 offset
                              Size = int32 size }
                        idx <- idx + 1
            if acc.Count = 0 then Error Truncated
            else Ok(List.ofSeq acc, pos)

    /// Extract the raw payload bytes for one entry (ZLIB wrapping intact).
    let readEntry (buf: byte[]) (entry: Entry) : byte[] =
        let start = int entry.Offset
        let mutable size = int entry.Size
        if start < 0 || start >= buf.Length then [||]
        else
            if start + size > buf.Length then size <- buf.Length - start
            let slice = Array.zeroCreate<byte> size
            Array.Copy(buf, start, slice, 0, size)
            slice
